tapestry-tool/tapestry-wp
# 1196 - Fix import/export bugs and limitations; add import from zip

## Changes
- Moves import file validation and permissions filtering to the backend.
- Adds ability to export a Tapestry to a ZIP file, which will bundle the regular JSON export file with all **locally uploaded** media (videos, images, H5Ps) in the Tapestry. This method is suitable for moving an entire Tapestry to a different site.
- If there are WordPress post nodes in the Tapestry, a WXR file (the export format for WordPress posts) is downloaded alongside the JSON or ZIP file. Otherwise, only the export file is downloaded. 
- Adds ability to import an exported Tapestry ZIP. This **re-creates** all media referenced by the Tapestry on the destination site.
- Expands the ImportChangelog component to display import warnings and information to the user.
- To import a Tapestry with WordPress posts, the user should import the posts before importing the Tapestry that uses them, and the WordPress Post nodes in the Tapestry will automatically link to the new posts.
- Splits Export and Duplicate functionalities in Settings modal since Export has expanded.
- Adds a handler to Cypress allowing it to print debug logs. Not directly applicable to this PR, but may help with debugging in the future.

**Bug fixes:**
- Fixes bug that prevents importing because `group` property does not exist
- Fixes bug when importing Answer nodes (since node IDs change during import, the `activityId` field should be updated to the new node ID)
- Fixes bug where imported Tapestries do not have a title (described at https://github.com/tapestry-tool/tapestry-wp/pull/1174#issuecomment-1171773108)

## Change details
Export details:
- Nodes with problems in the media (e.g. an image wasn't found) are exported as-is, but a warning is logged. The same goes for media in the Tapestry settings (e.g. the background image). At the end, all warnings are displayed to the user.
- Each export creates a temporary ZIP file in the `wp-content/uploads/tapestry/export` folder (plus a file with the same name and no extension). The URL of this ZIP is passed back to the client to be downloaded. A daily job should automatically delete export files older than 24 hours.

ZIP file details:
- The exported ZIP file has a flat structure (no root directory).
- The ZIP file will always contain a `tapestry.json`, which is the regular JSON export file (with some minor differences, described below)
- Image and video files are saved in the ZIP with the file names at the end of their URLs. (e.g. `http://localhost:8888/wp-content/uploads/video.mp4` becomes `video.mp4`)
- H5Ps are saved in the ZIP as the corresponding export file under `wp-content/uploads/h5p/exports` (e.g. `interactive-video-35.h5p`)
- Images added via Link Preview or videos added via URL (i.e. anything not a local upload) are not saved.
- Thumbnail images are saved in their original size, not the variants generated by WordPress 
- Images inside Activity nodes are only saved if the answer type and `useImages` are enabled.
- Inside tapestry.json, the URLs to local uploads are replaced by their file names inside the ZIP
     - e.g. `http://localhost:8888/wp-admin/admin-ajax.php?action=h5p_embed&id=3` -> `my-quiz-3.h5p`
     - e.g. `http://localhost:8888/wp-content/uploads/thumbnail.jpeg` -> `thumbnail.jpeg`

Import details:
- Any nodes that failed to import (e.g. a referenced file didn't exist) are still added, but a warning is logged. The same goes for imported Settings. At the end, all warnings are displayed to the user.
- Each import creates temporary directories under the `wp-content/uploads/tapestry/import` folders for unzipping the ZIP file. These should be deleted immediately after the import. (If they are not, something is wrong.)
- Additional sizes are generated for thumbnail images, but not for other images.
- H5P caches are only rebuilt if the imported Tapestry includes H5P files and the H5P files were uploaded successfully. Rebuilding the H5P cache is time-consuming, so we want to avoid it if possible.

## Screenshot
New Export / Duplicate modal:

<img width="507" alt="image" src="https://user-images.githubusercontent.com/8981258/196546508-861a93b9-e008-4678-a15d-de7ea639fa99.png">
<img width="508" alt="image" src="https://user-images.githubusercontent.com/8981258/196546621-f1c1a434-1c30-44aa-b55d-9d06ed0f7707.png">

Import (with warnings):

https://user-images.githubusercontent.com/44929104/178632980-abc03476-9975-487e-b494-5d11bf05ee77.mp4

Export (with warnings):

<img width="507" alt="image" src="https://user-images.githubusercontent.com/8981258/196546778-57942d8a-ebed-46fc-a205-e48e32670eb1.png">

Video with outdated UI showing export with warnings: https://user-images.githubusercontent.com/44929104/178630936-3a55c808-8b0f-4d04-a2fd-715d8c95b2a8.mp4

## Testing
**To test importing WordPress post nodes**: Since we only have 1 site to test on, we have to emulate the WordPress posts being recreated.
1. Export the Tapestry
2. Open the downloaded XML file. Search for all `<wp:post_id>` tags and change the numbers to something different.
3. Delete the exported posts from your local site. Make sure to also delete them from the Trash, since WordPress will warn you that the posts still exist otherwise.
4. Go to Tools -> Import. Install the "WordPress" importer if you haven't already, and click "Run Importer". Follow the instructions to import the WordPress file.
5. Import the Tapestry ZIP file. Verify that the WordPress post nodes refer to the newly created posts.

## Issue Linkage
Closes #1196 

## PR Dependency
Depends on #1188
 
## Automated Testing
Added tests to cover several basic cases:
- Several kinds of invalid ZIP files
- Importing a ZIP file containing images and videos, with no warnings
- Importing a ZIP file containing images and videos, with some warnings
- Exporting a ZIP file
- Exporting a tapestry with WordPress posts
- Exporting a ZIP file with warnings
- Importing a ZIP file that exported with warnings

Importing H5Ps and WordPress posts must be manually tested.
